File: julia/dopri.jl
export dopri, benchmark_dopri

function benchmark_dopri(times::Int)
    y = [8.59072560e+02, -4.13720368e+03, 5.29556871e+03, 7.37289205e+00, 2.08223573e+00, 4.39999794e-01]
    y0 = copy(y)
    mu = 3.986004418e5
    el = elements(y[1:3], y[4:6], mu)
    rpar = collect(mu)
    x = 0.0
    xend = period(el[1], mu)
    worst = -Inf
    best = Inf
    total = 0.0
    dopri!(gravity!, x, y, xend, rpar)
    for i = 1:times
        gc_enable(false)
        t = @elapsed dopri!(gravity!, x, y, xend, rpar)
        gc_enable(true)
        y = copy(y0)
        x = 0.0
        if t > worst
            worst = t
        end
        if t < best
            best = t
        end
        total += t
    end
    println("[$(total/times),$best,$worst]")
end

function gravity!(_n::Ptr{Cint}, _x::Ptr{Cdouble}, _y::Ptr{Cdouble}, _f::Ptr{Cdouble},
    _rpar::Ptr{Cdouble}, _ipar::Ptr{Cint})
    n = unsafe_load(_n, 1)
    t = unsafe_load(_x, 1)
    rpar = unsafe_load(_rpar, 1)
    y = pointer_to_array(_y, n)
    f = pointer_to_array(_f, n)

    r = sqrt(y[1]*y[1]+y[2]*y[2]+y[3]*y[3])
    r3 = r*r*r
    f[1] = y[4]
    f[2] = y[5]
    f[3] = y[6]
    f[4] = -rpar[1] * y[1] / r3
    f[5] = -rpar[1] * y[2] / r3
    f[6] = -rpar[1] * y[3] / r3
    return nothing
end

function _solout(_nr::Ptr{Cint}, _xold::Ptr{Cdouble}, _x::Ptr{Cdouble},
    _y::Ptr{Cdouble}, _n::Ptr{Cint}, _con::Ptr{Cdouble}, _icomp::Ptr{Cint},
    _nd::Ptr{Cint}, _rpar::Ptr{Cdouble}, _ipar::Ptr{Cint}, _irtrn::Ptr{Cint},
    _xout::Ptr{Cdouble})
    return nothing
end

cfcn = cfunction(gravity!, Void, (Ptr{Cint}, Ptr{Cdouble}, Ptr{Cdouble}, Ptr{Cdouble}, Ptr{Cdouble}, Ptr{Cint}))
csolout = cfunction(_solout, Void, (Ptr{Cint}, Ptr{Cdouble}, Ptr{Cdouble},
    Ptr{Cdouble}, Ptr{Cint}, Ptr{Cdouble}, Ptr{Cint},
    Ptr{Cint}, Ptr{Cdouble}, Ptr{Cint}, Ptr{Cint},
    Ptr{Cdouble}))

function dopri!(func::Function, x::Float64, y::Vector{Float64}, xend::Float64,
    rpar::Vector{Float64}=Float64[], reltol::Float64=1e-6, abstol::Float64=1e-8)
    n = length(y)
    lwork = 11*n + 8*n + 21
    liwork = n + 21
    work = zeros(Cdouble, lwork)
    iwork = zeros(Cint, liwork)
    ipar = Cint[]
    iout = 0
    idid = 0
    itol = 0
    rtol = collect(reltol)
    atol = collect(abstol)
    ccall((:c_dop853, :libdopri), Void, (Ptr{Cint}, Ptr{Void}, Ptr{Cdouble},
    Ptr{Cdouble}, Ptr{Cdouble}, Ptr{Cdouble}, Ptr{Cdouble},
    Ptr{Cint}, Ptr{Void}, Ptr{Cint}, Ptr{Cdouble},
    Ptr{Cint}, Ptr{Cint}, Ptr{Cint}, Ptr{Cdouble},
    Ptr{Cint}, Ptr{Cint}),
    &n, cfcn, &x, y, &xend, rtol, atol, &itol, csolout, &iout,
    work, &lwork, iwork, &liwork, rpar, ipar, &idid)
end

File: lib/dopri/dopri.f90
module dopri

use iso_c_binding, only: c_double, c_int, c_funptr, c_f_procpointer, c_associated, c_loc

implicit none

abstract interface
    subroutine dopfcn(n, x, y, f, rpar, ipar)
        integer, intent(inout) :: n
        double precision, intent(inout) :: x
        double precision, dimension(n), intent(inout) :: y
        double precision, dimension(:), intent(inout) :: rpar
        double precision, dimension(n), intent(inout) :: f
        integer, dimension(:), intent(inout) :: ipar
    end subroutine dopfcn
    subroutine dopsolout(nr, xold, x, y, n, con, icomp,&
            nd, rpar, ipar, irtrn, xout)
        integer, intent(inout) :: nr
        double precision, intent(inout) :: xold
        double precision, intent(inout) :: x
        double precision, dimension(n), intent(inout) :: y
        integer, intent(inout) :: n
        double precision, dimension(8*nd), intent(inout) :: con
        integer, dimension(nd), intent(inout) :: icomp
        integer, intent(inout) :: nd
        double precision, dimension(:), intent(inout) :: rpar
        integer, dimension(:), intent(inout) :: ipar
        integer, intent(inout) :: irtrn
        double precision, intent(inout) :: xout
    end subroutine dopsolout
end interface

interface
    subroutine dop853(n, fcn, x, y, xend, rtol, atol,&
            itol, solout, iout, work, lwork, iwork,&
            liwork, rpar, ipar, idid)
        procedure(dopfcn), pointer :: fcn
        procedure(dopsolout), pointer :: solout
        integer, intent(inout) :: n
        integer, intent(inout) :: itol
        integer, intent(inout) :: iout
        integer, intent(inout) :: lwork
        integer, intent(inout) :: liwork
        integer, dimension(:),intent(inout) :: ipar
        integer, intent(inout) :: idid
        double precision, intent(inout) :: xend
        double precision, dimension(n), intent(inout) :: rtol
        double precision, dimension(n), intent(inout) :: atol
        double precision, dimension(:), intent(inout) :: rpar
        double precision, dimension(lwork), intent(inout) :: work
        integer, dimension(liwork), intent(inout) :: iwork
        double precision, intent(inout) :: x
        double precision, dimension(n), intent(inout) :: y
    end subroutine dop853

    double precision function contd8(ii, x, con, icomp, nd)
        integer, intent(inout) :: ii
        double precision, intent(inout) :: x
        double precision, dimension(8*nd), intent(inout) :: con
        integer, dimension(nd), intent(inout) :: icomp
        integer, intent(inout) :: nd
    end function contd8
end interface

abstract interface
    subroutine c_fcn(n, x, y, f, rpar, ipar)
        import :: c_int
        import :: c_double
        integer(c_int), intent(inout) :: n
        integer(c_int), dimension(:),intent(inout) :: ipar
        real(c_double), intent(inout) :: x
        real(c_double), dimension(n), intent(inout) :: y
        real(c_double), dimension(:), intent(inout) :: rpar
        real(c_double), dimension(n), intent(inout) :: f
    end subroutine c_fcn
    subroutine c_solout(nr, xold, x, y, n, con, icomp,&
            nd, rpar, ipar, irtrn, xout)
        import :: c_int
        import :: c_double
        integer(c_int), intent(inout) :: n
        integer(c_int), intent(inout) :: nr
        integer(c_int), intent(inout) :: nd
        integer(c_int), intent(inout) :: irtrn
        integer(c_int), dimension(:), intent(inout) :: ipar
        integer(c_int), dimension(nd), intent(inout) :: icomp
        real(c_double), intent(inout) :: xold
        real(c_double), intent(inout) :: x
        real(c_double), dimension(n), intent(inout) :: y
        real(c_double), dimension(8*nd), intent(inout) :: con
        real(c_double), dimension(:), intent(inout) :: rpar
        real(c_double), intent(inout) :: xout
    end subroutine c_solout
end interface

contains

subroutine c_dop853(n, cfcn, x, y, xend, rtol, atol,&
        itol, csolout, iout, work, lwork, iwork,&
        liwork, rpar, ipar, idid) bind(c)
    integer(c_int), intent(inout) :: n
    type(c_funptr), intent(in), value :: cfcn
    real(c_double), intent(inout) :: x
    real(c_double), dimension(n), intent(inout) :: y
    real(c_double), intent(inout) :: xend
    real(c_double), dimension(n), intent(inout) :: rtol
    real(c_double), dimension(n), intent(inout) :: atol
    integer(c_int), intent(inout) :: itol
    type(c_funptr), intent(in), value :: csolout
    integer(c_int), intent(inout) :: iout
    real(c_double), dimension(lwork), intent(inout) :: work
    integer(c_int), intent(inout) :: lwork
    integer(c_int), dimension(liwork), intent(inout) :: iwork
    integer(c_int), intent(inout) :: liwork
    real(c_double), dimension(:), intent(inout) :: rpar
    integer(c_int), dimension(:),intent(inout) :: ipar
    integer(c_int), intent(inout) :: idid

    procedure(c_fcn), pointer :: fcn
    procedure(c_solout), pointer :: solout

    call c_f_procpointer(cfcn, fcn)
    call c_f_procpointer(csolout, solout)

    print *, c_associated(cfcn)
    print *, loc(cfcn)
    print *, associated(fcn)
    print *, loc(fcn)

    call dop853(n, fcn, x, y, xend, rtol, atol,&
            itol, solout, iout, work, lwork, iwork,&
            liwork, rpar, ipar, idid)
end subroutine c_dop853

function c_contd8(ii, x, con, icomp, nd) result(ret) bind(c)
    real(c_double) :: ret
    integer(c_int), intent(inout) :: ii
    real(c_double), intent(inout) :: x
    real(c_double), dimension(8*nd), intent(inout) :: con
    integer(c_int), dimension(nd), intent(inout) :: icomp
    integer(c_int), intent(inout) :: nd
    ret = contd8(ii, x, con, icomp, nd)
end function c_contd8

end module
